### PR TITLE
enh(js/ts) improve CLASS_REFERENCE

### LINKED ISSUE #3403: (TypeScript) Camel-case type parameters: only first component is highlighted
**Describe the issue**

In the type parameter "OutT", the "Out" is highlighted differently from the "T".

**Which language seems to have the issue?**

`typescript`

**Are you using `highlight` or `highlightAuto`?**

Not sure.  I noticed this issue on StackOverflow and reproduced with your JSFiddle template.

**Sample Code to Reproduce**

https://jsfiddle.net/650rLwyd/

Screenshot:

Screen Shot 2021-11-20 at 15 30 25

**Expected behavior**

GitHub's highlighting in markdown code fences looks better:

```ts
type ParseFunc<OutT, InT> = (val: InT) => OutT;

declare const parseString: ParseFunc<string, unknown>;

declare const makeParseObject: <FPs extends {[s: string]: ParseFunc<any, any>}> (fieldParsers: FPs)
    => ParseFunc<ParseObjectResult<FPs>, unknown>;
type ParseObjectResult<FPs extends {[s: string]: ParseFunc<any, unknown>}> =
    {[F in keyof FPs]: ParseResult<FPs[F]>}
type ParseResult<P> = P extends ParseFunc<infer OutT, unknown> ? OutT : never;
```

Screenshot, just in case it changes:

Screen Shot 2021-11-20 at 15 27 32

**Additional context**

### PR DESCRIPTION
**Pull request by joshgoebel**
Resolves #3403.

### Changes

- improves CLASS_NAME_REFERNCE rule to catch more edge cases

### Checklist
- [x] Added markup tests, or they don't apply here because...
- [x] Updated the changelog at `CHANGES.md`

### CODE REVIEW

File: src/languages/typescript.js [hunk 1]
```diff
@@ -23,7 +23,8 @@ export default function(hljs) {
     "string",
     "object",
     "never",
-    "enum"
+    "enum",
+    "unknown"
```
Comment (joshgoebel) [inline src/languages/typescript.js:27]:
```suggestion
    "enum"
```

File: src/languages/typescript.js [hunk 2]
```diff
@@ -61,7 +62,9 @@ export default function(hljs) {
     "implements",
     "declare",
     "abstract",
-    "readonly"
+    "readonly",
+    "infer",
+    "keyof"
```
Comment (joshgoebel) [inline src/languages/typescript.js:66]:
```suggestion
    "readonly"
```

**Review by allejo — CHANGES_REQUESTED** [review]
Missing the edge case of a class matching the "CSSFactory" rule and ending with an uppercase letter; e.g. `HTTPResponseT` or `IResponseTsS`

image

**Review by allejo — APPROVED** [review]
lgtm :+1:
